### PR TITLE
Add a check that primary and secondary use `custom_format` or not

### LINKED ISSUE #1150: Not to warn about secondary plugin type if standard format is used in both of primary and secondary
Now, Fluentd shows warning if secondary plugin type is different from primary plugin type. But it is nonsense in some cases in Fluentd v0.14. This propose to suppress it if both of primary and secondary use "standard" format, instead of implementing `format` method by themselves.

Fluentd v0.12 requires buffered output plugins to implement `format` method by themselves. It means that different plugin types have different formatting of buffer chunks.
Secondary plugin uses buffer chunks formatted by primary plugins. So different plugin types between primary and secondary should causes unexpected errors or failures to operate buffer chunks.

But Fluentd v0.14 has "standard" format, so different plugin types can share the way to format buffer chunks by using standard format.
There is no need to warn about type difference between primary and secondary if both these use standard format.
It can be implemented by checking plugins implement `format` method by themselves or not.

### PR DESCRIPTION
**Pull request by ganmacs**
Not to warn about secondary plugin type if "standard" format is used in both of primary and secondary.

This PR closes #1150.

### CODE REVIEW

File: test/test_output.rb [hunk 1]
```diff
@@ -95,12 +95,37 @@ def write(chunk)
...
     def test_secondary
-      d = create_driver(CONFIG + %[
+      d = Fluent::Test::BufferedOutputTestDriver.new(Fluent::BufferedOutput) do
+        def write(chunk)
+          chunk.read
+        end
+      end
+
+      mock(d.instance.log).warn("secondary type should be same with primary one",
+                                { :primary=>d.instance.class.to_s, :secondary=>"Fluent::Plugin::Test2Output" })
```
Comment (tagomoris) [inline test/test_output.rb]: Use json style hash literal instead of rocket operator.

**Comment by tagomoris** [comment]
`test/test_output` is tests for traditional (v0.12) output class. Adding tests in that file is not bad idea, but we need to write tests for current `Fluent::Plugin::Output` in `test/plugin/test_output.rb`.

File: test/test_output.rb (hunk 1 above)
Comment (ganmacs) [inline test/test_output.rb]: fixed.
https://github.com/fluent/fluentd/pull/1153/files#diff-cc6de9d6863a116b8fe4eb5d954a31aeR105

**Comment by ganmacs** [comment]
@tagomoris 

> test/test_output is tests for traditional (v0.12) output class.

I don't know that 😨 
I add a new test for `test/plugin/output.rb`

https://github.com/fluent/fluentd/pull/1153/commits/e4e2204dfaca48b757fd386915f4b5075195003a

File: test/plugin/test_output.rb [hunk 2]
```diff
@@ -477,6 +488,36 @@ def waiting(seconds)
 
       i.stop; i.before_shutdown; i.shutdown; i.after_shutdown; i.close; i.terminate
     end
+
+    test "Warn if primary type differents from secondary type and either primary or secondary has custom_format" do
```
Comment (tagomoris) [inline test/plugin/test_output.rb]: "is different from"

File: test/plugin/test_output.rb [hunk 3]
```diff
@@ -477,6 +488,36 @@ def waiting(seconds)
...
+                                { primary: o.class.to_s, secondary: "Fluent::Plugin::TestOutput" }).never
+
+      o.configure(config_element('ROOT','',{'name' => "cool2"},
+                                 [config_element('secondary','',{'@type'=>'test', 'name' => "cool"}),
+                                  config_element('buffer','',{'@type'=>'memory'})]
+                                ))
+      assert_not_nil o.instance_variable_get(:@secondary)
+    end
+
+    test "Not to warn if primary type differents from secondary type and both do'nt have custom_format" do
```
Comment (tagomoris) [inline test/plugin/test_output.rb]: "don't"

File: test/plugin/test_output.rb (hunk 2 above)
Comment (ganmacs) [inline test/plugin/test_output.rb]: fixed.
https://github.com/fluent/fluentd/pull/1153/commits/af8056f1addee4e0b1bd30dd50da0c4e8042f55b#diff-ed96b0313fe0cd126f17ce6909e72f1aR492

File: test/plugin/test_output.rb (hunk 3 above)
Comment (ganmacs) [inline test/plugin/test_output.rb]: fixed.
https://github.com/fluent/fluentd/pull/1153/commits/af8056f1addee4e0b1bd30dd50da0c4e8042f55b#diff-ed96b0313fe0cd126f17ce6909e72f1aR501

File: test/plugin/test_output.rb [hunk 4]
```diff
@@ -477,6 +488,36 @@ def waiting(seconds)
...
+    test "Warn if primary type is different from secondary type and either primary or secondary has custom_format" do
+      o = create_output(:buffered)
+      mock(o.log).warn("secondary type should be same with primary one",
+                                { primary: o.class.to_s, secondary: "Fluent::Plugin::TestOutput" })
+
+      o.configure(config_element('ROOT','',{},[config_element('secondary','',{'@type'=>'test', 'name' => "cool"})]))
+      assert_not_nil o.instance_variable_get(:@secondary)
+    end
+
+    test "don't warn if primary type is same as econdary type" do
```
Comment (tagomoris) [inline test/plugin/test_output.rb]: "econdary" -> "secondary"

File: test/plugin/test_output.rb [hunk 5]
```diff
@@ -477,6 +488,36 @@ def waiting(seconds)
...
+                                { primary: o.class.to_s, secondary: "Fluent::Plugin::TestOutput" }).never
+
+      o.configure(config_element('ROOT','',{'name' => "cool2"},
+                                 [config_element('secondary','',{'@type'=>'test', 'name' => "cool"}),
+                                  config_element('buffer','',{'@type'=>'memory'})]
+                                ))
+      assert_not_nil o.instance_variable_get(:@secondary)
+    end
+
+    test "don't warn if primary type is different from secondary type and both do'nt have custom_format" do
```
Comment (tagomoris) [inline test/plugin/test_output.rb]: "do'nt" -> "don't"

File: test/plugin/test_output.rb (hunk 4 above)
Comment (ganmacs) [inline test/plugin/test_output.rb]: Fixed
https://github.com/fluent/fluentd/pull/1153/files#diff-ed96b0313fe0cd126f17ce6909e72f1aR501

File: test/plugin/test_output.rb (hunk 5 above)
Comment (ganmacs) [inline test/plugin/test_output.rb]: Fixed.
https://github.com/fluent/fluentd/pull/1153/files#diff-ed96b0313fe0cd126f17ce6909e72f1aR513

**Comment by tagomoris** [comment]
LGTM. Thank you for contribution!

**Pull request merged**